File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.cpp
class Solution {
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        const int MOD = 1000000007;
        vector<vector<vector<int>>> memo(
            m, vector<vector<int>>(n, vector<int>(maxMove + 1, -1))
        );

        function<int(int, int, int)> dfs = [&](int r, int c, int movesLeft) -> int {
            if (r < 0 || r >= m || c < 0 || c >= n) return 1;
            if (movesLeft == 0) return 0;
            if (memo[r][c][movesLeft] != -1) return memo[r][c][movesLeft];

            long long ans = 0;
            ans += dfs(r + 1, c, movesLeft - 1);
            ans += dfs(r - 1, c, movesLeft - 1);
            ans += dfs(r, c + 1, movesLeft - 1);
            ans += dfs(r, c - 1, movesLeft - 1);

            return memo[r][c][movesLeft] = ans % MOD;
        };

        return dfs(startRow, startColumn, maxMove);
    }
};

File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.py
"""
dfs(r, c, movesLeft)
=
dfs(r+1, c, movesLeft-1)
+ dfs(r-1, c, movesLeft-1)
+ dfs(r, c+1, movesLeft-1)
+ dfs(r, c-1, movesLeft-1)

memoization
= dp[r][c][k]: (r, c)에서 movesLeft번 남았을 때 밖으로 나가는 경로 수
"""
from functools import lru_cache

class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        MOD = 10**9 + 7

        @lru_cache(None)
        def dfs(r, c, movesLeft):
            if not (0 <= r < m and 0 <= c < n):
                return 1
            if movesLeft == 0:
                return 0

            return (
                dfs(r + 1, c, movesLeft - 1)
                + dfs(r - 1, c, movesLeft - 1)
                + dfs(r, c + 1, movesLeft - 1)
                + dfs(r, c - 1, movesLeft - 1)
            ) % MOD

        return dfs(startRow, startColumn, maxMove)
        